src/orchestration/orchestration.cpp

```diff
@@ -0,0 +1,911 @@
+    ERR_FAIL_COND(p_node.is_null());
+    ERR_FAIL_COND(_nodes.has(p_node->get_id()));
+
+    // Validate the node details
+    p_node->_orchestration = this;
+    p_node->_validate_input_default_values();
+
+    // Register the node with the script
+    _nodes[p_node->get_id()] = p_node;
+
+    // Register the node with the graph
+    p_graph->add_node(p_node);
+}
+
+void Orchestration::remove_node(int p_node_id)
+{
+    ERR_FAIL_COND_MSG(_has_instances(), "Cannot remove node, instances exist.");
+    ERR_FAIL_COND(!_nodes.has(p_node_id));
+
+    const Ref<OScriptNode> node = _nodes[p_node_id];
+    node->pre_remove();
+
+    // Check whether the node represents a function and if so, remove the function
+    for (const KeyValue<StringName, Ref<OScriptFunction>>& E : _functions)
+    {
+        if (E.value->get_owning_node_id() == p_node_id)
+        {
+            _functions.erase(E.key);
+            break;
+        }
+    }
+
+    for (const Ref<OScriptNodePin>& pin : node->get_all_pins())
+        pin->unlink_all(true);
+
+    List<OScriptConnection> removals;
+    for (const OScriptConnection& connection : _connections)
+    {
+        if (connection.is_linked_to(p_node_id))
+            removals.push_back(connection);
+    }
+
+    if (!removals.is_empty())
+    {
+        ERR_PRINT("Node still has remaining connects, cleaning them up");
+        while (!removals.is_empty())
+        {
+            _connections.erase(removals.front()->get());
+            removals.pop_front();
+        }
+
+    }
+
+    for (const KeyValue<StringName, Ref<OScriptGraph>>& E : _graphs)
+        E.value->remove_node(node);
+
+    _nodes.erase(p_node_id);
+}
+
+Ref<OScriptNode> Orchestration::get_node(int p_node_id) const
+{
+    ERR_FAIL_COND_V_MSG(!_nodes.has(p_node_id), nullptr, "No node exists with the specified ID: " + itos(p_node_id));
+    return _nodes[p_node_id];
+}
+
+Vector<Ref<OScriptNode>> Orchestration::get_nodes() const
+{
+    Vector<Ref<OScriptNode>> results;
+    for (const KeyValue<int, Ref<OScriptNode>>& E : _nodes)
+        results.push_back(E.value);
+    return results;
+}
+
+const RBSet<OScriptConnection>& Orchestration::get_connections() const
+{
+    return _connections;
+}
+
+void Orchestration::connect_nodes(int p_source_id, int p_source_port, int p_target_id, int p_target_port)
+{
+    ERR_FAIL_COND_MSG(_has_instances(), "Cannot connect nodes, instances exist.");
+
+    OScriptConnection connection;
+    connection.from_node = p_source_id;
+    connection.from_port = p_source_port;
+    connection.to_node = p_target_id;
+    connection.to_port = p_target_port;
+
+    ERR_FAIL_COND_MSG(_connections.has(connection), "A connection already exists: " + connection.to_string());
+    _connections.insert(connection);
+
+    _self->emit_signal("connections_changed", "connect_nodes");
+}
+
+void Orchestration::disconnect_nodes(int p_source_id, int p_source_port, int p_target_id, int p_target_port)
```

Comment:
Could `disconnect_nodes` only exist publically on `OScriptGraph` and then it has friend access to `Orchestration`?

Reply:
There is an order of operations issue with `OScriptNodePin`, that we will need to fix before this can be done.